Display Checkout block with a notice when adding a product sold individually that was already in the cart

Aljullu opened the pull request:
Fixes #2670.

Some changes included in this PR:
* Notices generated before running the Checkout block logic are printed before. This ensures that when CartController [calls](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/src/StoreApi/Utilities/CartController.php#L244) `NoticeHandler::convert_notices_to_exceptions()` there are not old notices.
* Renamed `CheckoutError` to `CheckoutOrderError` so it matches its directory name. 2278519f97114a16bd7b940e8a0661e5e3a63ea9
* Replaced two strings to match the same strings in Core, so we reduce the effort to localize WC. 957152137baac567e994546a1deae848c5b9bccb

Tagging @mikejolley for review since you worked on this and will probably know if this approach makes sense or there is a better way I missed.

### Screenshots

![imatge](https://user-images.githubusercontent.com/3616980/87408529-e33de380-c5c3-11ea-82d4-13b89b7e88d7.png)

### How to test the changes in this Pull Request:

There are steps to reproduce in #2670, but I think they can be simplified to:

1. Edit a product so it can only be added once to the Cart.
![imatge](https://user-images.githubusercontent.com/3616980/87408703-27c97f00-c5c4-11ea-9ebb-c2dfdcc5472e.png)
2. Go to `https://one.wordpress.test/checkout/?add-to-cart=19` (replacing `one.wordpress.test` with your test site URL and `19` with the ID of the product you modified).
3. Verify the Checkout block appears correctly and the error message appears as a notice.

### Changelog

> Fixed a bug that was causing the Checkout block to break when adding a product sold individually that was already in the cart

mikejolley commented:
@Aljullu just so I understand this, is the notice thats showing in this example coming from code unrelated to blocks? i.e. core error notices from the query string?

Would another way to solve this be to clear notices before running the checkout logic and converting notices to exceptions?

Aljullu commented:
> @Aljullu just so I understand this, is the notice thats showing in this example coming from code unrelated to blocks? i.e. core error notices from the query string?

Correct, that error is [thrown by WC Core](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-cart.php#L1032-L1035) when a product is added to the Cart.

The issue we are facing in #2670 is that checkout data is not returned by `/wc/store/checkout` if accessing the Checkout page with the `add-to-cart` argument in the URL. Instead, the error generated by Core is returned.

> Would another way to solve this be to clear notices before running the checkout logic and converting notices to exceptions?

It might, but this part of the codebase is totally new to me, so I trust you more. :smile:  From my understading we are already converting notices to exceptions, right? Clearing notices before running the checkout logic would fix the issue of the Checkout data not being available, but we would not be displaying any feedback to the user that the product was not added to the cart.

mikejolley commented:
It's tricky because we don't want non-checkout notices to pollute the checkout. I also don't think it's desirable to this query string to be passed to the API requests (is that what is happening?)—I only see usage of the exception conversion in the API itself. Are we somehow passing the add-to-cart string to the API requests? That might be a bug? 

I think the purpose of `convert_notices_to_exceptions` in this context is to only convert the notices thrown by it's internal logic. The notices that were there before might need discarding.

haszari commented:
Edited description to fix issue link (was linking to #2679).

> That notice has HTML tags and their content stripped. WC Core throws some errors with links, but that was not playing well with the Notice component (it has the possibility to render HTML, but it's marked as unstable).

Thanks for digging into this 👍 

We could use a link in some cart error notices, e.g. "please log in to your existing account" - hit this in #2851

Aljullu commented:
@mikejolley 

> It's tricky because we don't want non-checkout notices to pollute the checkout. I also don't think it's desirable to this query string to be passed to the API requests (is that what is happening?)—I only see usage of the exception conversion in the API itself. Are we somehow passing the add-to-cart string to the API requests? That might be a bug?

I don't think so, the initial contents of the Checkout page are [hydrated directly via PHP](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/src/BlockTypes/Checkout.php#L193) and we don't pass any `params` to `get_endpoint_data()`, so it shouldn't have the `add-to-cart` string, no?

> I think the purpose of `convert_notices_to_exceptions` in this context is to only convert the notices thrown by it's internal logic. The notices that were there before might need discarding.

If we plainly discarded them, there would be no feedback for the user that the product was not added to their cart. What I was trying to achieve with this PR is that notices thrown by internal logic are not converted to exceptions but are not discarded either, instead they would be passed in a new `notice` property.

Aljullu commented:
@haszari 

> We could use a link in some cart error notices, e.g. "please log in to your existing account" - hit this in #2851

That would mean we need to define own custom message, no? The issue I see from that (if I'm understanding it right) is that all this logic is currently in WC Core, so Blocks doesn't really know what error happened.

mikejolley commented:
> I don't think so, the initial contents of the Checkout page are hydrated directly via PHP and we don't pass any params to get_endpoint_data(), so it shouldn't have the add-to-cart string, no?

Good spot. It doesn't totally make sense to me where these notices are coming from in that case, I guess added just before that hydration occurs. If that is indeed the case, we could hydrate and clear existing notices to another property if that's easier. I defer to your judgement :)

haszari commented:
> > We could use a link in some cart error notices, e.g. "please log in to your existing account" - hit this in #2851

> That would mean we need to define own custom message, no? The issue I see from that (if I'm understanding it right) is that all this logic is currently in WC Core, so Blocks doesn't really know what error happened.

This comment should not block this PR at all – I was just mentioning a feature request / need that's possibly related. That being that error notices might need links. Specifically, some existing WC core error notices [do contain links](https://github.com/woocommerce/woocommerce/blob/e736ac4145d9aceb3574922d01824ede546a07f3/includes/wc-user-functions.php#L47), and we might strip the links out when we render the notices (I'm [stripping the links](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2851/files#diff-e82bdb511fe4e8a787289e47fef92f88R375) in #2851).

Here's a link to the related comment thread: https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2851#pullrequestreview-453749668

But this is all pretty tangential (unrelated) to this PR and shouldn't block it, hopefully this clarifies things :)

Aljullu commented:
Updated this PR so it should be ready for review again. I updated the first comment with details on what's changed.

> This comment should not block this PR at all – I was just mentioning a feature request / need that's possibly related. That being that error notices might need links. Specifically, some existing WC core error notices do contain links, and we might strip the links out when we render the notices (I'm stripping the links in #2851).

Ah, yes. The [`Notice`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/notice) component allows having `actions`, but I'm not sure if that's well suited for our use-cases or we would need another solution.

haszari commented on docs/block-client-apis/checkout/checkout-api.md:
```diff
@@ -106,6 +106,7 @@ Via `useCheckoutContext`, the following are exposed:
 -   `orderId`: The order id for the order attached to the current checkout.
 -   `isCart`: This is true if the cart is being viewed. Note: usage of `CheckoutProvider` will automatically set this to false. There is also a `CartProvider` that wraps children in the `ShippingDataProvider` and exposes the same api as checkout context. The `CartProvider` automatically sets `isCart` to true. This allows components that implement `useCheckoutContext` to use the same api in either ...
 -   `hasOrder`: This is true when orderId is truthy.
+-   `notices`: An array of notices or messages associated to the current checkout data. For example, errors from the server which are not blocking the purchase process but have to be shown to the user.
```
Since the notices are captured as part of hydration (only), maybe we can call them `captured_hydrate_notices`. Again, this is not very clear at all 😄

haszari commented:
I grabbed this and reviewed since Mike may be out for a bit 🏖️ 👾

haszari commented:
Tangentially related enhancement issue - ensure blocks notices support actions :) https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2985

Aljullu commented on src/BlockTypes/Checkout.php:
```diff
@@ -179,18 +179,42 @@ protected function hydrate_customer_payment_methods( AssetDataRegistry $data_reg
+	 * @param string $text Text to remove tags from.
+	 */
+	protected function strip_tags_content( $text ) {
+		return preg_replace( '@<(\w+)\b.*?>.*?</\1>@si', '', $text );
```
In the end I changed how this is done (80568e9f1786cb947698dcc2e71614ff9d8952e8). Now, HTML is stripped in the frontend using a library WC Admin seems to use too. In some way I think it's more correct to let the client decide whether to display HTML markup or not.

This raises the bundle size a bit, so I'm happy for other ideas.

Aljullu commented on src/BlockTypes/Checkout.php:
```diff
@@ -179,18 +179,42 @@ protected function hydrate_customer_payment_methods( AssetDataRegistry $data_reg
 	protected function hydrate_from_api( AssetDataRegistry $data_registry ) {
+		// Save previous notices so we can return them in a `notices` property to
+		// be used in the frontend.
+		$previous_notices = wc_get_notices( 'error' );
```
I removed this limitation. :slightly_smiling_face:

Aljullu commented on src/BlockTypes/Checkout.php:
```diff
@@ -179,18 +179,42 @@ protected function hydrate_customer_payment_methods( AssetDataRegistry $data_reg
 			add_filter( 'woocommerce_store_api_disable_nonce_check', '__return_true' );
-			$data_registry->add( 'checkoutData', WC()->api->get_endpoint_data( '/wc/store/checkout' ) );
+			$checkout_data = WC()->api->get_endpoint_data( '/wc/store/checkout' );
+			if ( is_array( $previous_notices ) && count( $previous_notices ) > 0 ) {
```
> If we do this for checkout, we should do it for cart block too (and consider other cases).

The Cart block shouldn't be affected. The issue happened because Checkout is calling `convert_notices_to_exceptions`, that means any notice would prevent the Checkout block from rendering and instead an error would be shown. Cart doesn't call that method so it's not affected and it shows notices like other pages.

Aljullu commented:
Love all the suggestions @haszari! I implemented most of them and I will answer some of your feedback.

> First up a disclaimer - I'm conflicted about this, it feels like a bit of an edge case, or more specifically a general bug, not specific to cart or checkout.

The issue is only reproducible in the Checkout block because it calls `convert_notices_to_exceptions`, so previously-generated notices where making the block to break. Other blocks shouldn't be affected. I wouldn't say it's an edge case, though, it has been [reported by a user](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2670) and it sound to me like something that needs to be fixed.

> Long term, I'm thinking we'll provide new ways for merchants to build a landing page with an Add to cart button, and that might be the cleanest "fix" for these kinds of problems. The best time to show the You can't add error is as close to the user action as possible, so an AJAX error would be ideal.

:100: However, there will still be the case of links in the form `?add_to_cart=19`, in that case the backend tries to add the product to the cart on page load, so there is no way to do it in an AJAX-way.

> Perhaps this PR is more of an interim solution, or polish to protect inappropriate errors from causing confusion in cart & checkout block.

Correct, I think WC should give users feedback as soon as possible without requiring a page reload. But that's not yet possible with notices generated in Core and there will still be cases in which notices are generated on page load, as described above. So we still need a system so previously-generated notices don't break our blocks.

> Can you think of any other error notices that might "leak" into cart & checkout in this way?

I don't, but that's just because my WC knowledge is limited. :slightly_smiling_face: I'm pretty sure there must be some other ways to trigger errors or notices right before the block logic is run.

---

I renamed several variables with your suggestions. In the API I named it `captured_notices`, (I skipped `hydrate` because, in theory, that endpoint could be used without initial hydration). In the JS-side I named it `hydratedNotices`, in this case the frontend doesn't care if they were captured or not, the only important thing is that they came with hydrated data.

This PR is ready for another look.

haszari commented on src/BlockTypes/Checkout.php:
```diff
@@ -179,18 +179,41 @@ protected function hydrate_customer_payment_methods( AssetDataRegistry $data_reg
 	 * @param AssetDataRegistry $data_registry Data registry instance.
 	 */
 	protected function hydrate_from_api( AssetDataRegistry $data_registry ) {
+		$captured_notices = $this->capture_and_clear_notices();
```
Do we want to capture notices every time `hydrate_from_api` is called, or should this be protected?

Aljullu commented:
Thanks for the in-depth review @haszari! After digging a bit more into this, it looks like (maybe) this could be fixed with just one line. :facepalm: Printing the notices before running the block logic was enough to ensure they are empty when reaching `NoticeHandler::convert_notices_to_exceptions()`. Can you take another look?

> Click Add to cart - cart button. Button link navigates to cart block page – no notice is displayed. Cart only contains 1x Belt.

I'm still unable to reproduce this. Let me know if you can still reproduce with the last changes in this pull.

senadir commented:
@Aljullu is this ready to be merged?

Aljullu commented:
> @Aljullu is this ready to be merged?

@senadir I changed the milestone to 3.3.0, there is not need to rush this PR IMO. This way I can do some testing with the steps shared by @haszari. Thanks for asking! :slightly_smiling_face:

Aljullu commented:
@haszari, thanks for writing the testing steps, I could reproduce what you mean.

Let me split my response in several parts:

### Notices not always appearing in Twenty Twenty.

This seems to be the expected behavior. Even though Storefront displays WC notices [on top of each page](https://github.com/woocommerce/storefront/blob/master/inc/woocommerce/storefront-woocommerce-template-hooks.php#L42), that's not the case for other themes like Twenty Twenty. In them, notices are only shown when WC Core prints them. That happens in [some occasions](https://github.com/woocommerce/woocommerce/blob/master/includes/wc-template-hooks.php#L303-L314) like when rendering the Cart and Checkout blocks, but not on every page load.

I think that's the cause of the confusing behavior you were experimenting: you were triggering the error several times, but since you were visiting a regular page, notices were not being printed but instead were 'accumulated'. Once you visited the Checkout shortcode, all notices appeared at once.

I'm not sure how this should affect blocks, I think there are three possibilities:

1. We always print notices when rendering a block.
2. Cart and Checkout block print notices, but all others don't.
3. Only the Checkout block prints notices and all other blocks let the theme handle that (this PR).

I'm leaning towards option 1 or 2. Thoughts? In any case, I would open an issue for that since I don't think that's blocking this PR.

### Should blocks use an action instead of directly calling `wc_print_notices()`?

In the link I posted above, you can see that pages and shortcodes in WC Core don't call `wc_print_notices()` directly, but instead [use actions](https://github.com/woocommerce/woocommerce/blob/master/includes/wc-template-hooks.php#L303-L314) letting extensions hook into the process. Should we go for that approach too? Something like:

```PHP
add_action( 'woocommerce_blocks_before_cart_block', 'woocommerce_output_all_notices', 10 );
add_action( 'woocommerce_blocks_before_checkout_block', 'woocommerce_output_all_notices', 10 );
```

This would mean adding a new extensibility point. While I think that would be useful for extensions, I'm a bit wary of adding actions before we have completely decided on a direction on how to implement extensibility in our blocks. So I would vote this is not blocking right now but something to keep in mind. cc @nerrad

### Visual glitches

I created a PR in core to fix several visual glitches: https://github.com/woocommerce/woocommerce/pull/27387.

haszari commented:
Excellent investigation and analysis @Aljullu ! It seems to me you've hit on an area that's inconsistent across themes, and it's not clear what the "correct" behaviour is, or what Woo expects. Also we should keep in mind that the issue that triggered this is somewhat of an edge case, so the behaviour differences we're seeing might actually be low-impact. 

That said, this seems like an opportunity to improve the robustness of Woo + themes 😄 (at some point - not an urgent issue IMO).

> This seems to be the expected behavior. Even though Storefront displays WC notices on top of each page, that's not the case for other themes like Twenty Twenty

Aha! That does explain things. I wonder if we (woo core) should advocate for more themes to take the Storefront approach, or implement it in core as default behaviour.

> I'm leaning towards option 1 or 2. Thoughts? In any case, I would open an issue for that since I don't think that's blocking this PR.

I agree this should not block this PR, and it's probably a wider (woo platform) discussion, as this is really a woo core issue that we're hitting in blocks. I could live with any those options you mention – let's see what others think. Option 3 seems the least "intrusive" to me; all options seem like a patch the symptom rather than addressing the cause.

> Should blocks use an action instead of directly calling wc_print_notices()?

I agree - my instinct is that we should avoid adding hooks (or triggering existing hooks) until it's we know it's essential and needed.

Aljullu commented:
I'm going to merge this as-is since it has already been approved. We can follow-up with printing (or not) notices in more blocks and extensibility in the future.
